Sources/Algorithms/Suffix.swift

```diff
@@ -27,13 +27,62 @@ extension BidirectionalCollection {
   public func suffix(
     while predicate: (Element) throws -> Bool
   ) rethrows -> SubSequence {
-    let start = startIndex
-    var result = endIndex
-    while result != start {
-      let previous = index(before: result)
-      guard try predicate(self[previous]) else { break }
-      result = previous
+    try self[startOfSuffix(while: predicate)...]
+  }
+}
+
+//===----------------------------------------------------------------------===//
+// endOfPrefix(while:)
+//===----------------------------------------------------------------------===//
+
+extension Collection {
+  /// Returns the exclusive upper bound of the prefix of elements that satisfy
+  /// the predicate.
+  ///
+  /// - Parameter predicate: A closure that takes an element of the collection
+  ///   as its argument and returns `true` if the element is part of the prefix
+  ///   or `false` if it is not. Once the predicate returns `false` it will not
+  ///   be called again.
+  ///
+  /// - Complexity: O(*n*), where *n* is the length of the collection.
+  @usableFromInline
+  internal func endOfPrefix(
+    while predicate: (Element) throws -> Bool
+  ) rethrows -> Index {
+    var index = startIndex
+    while try index != endIndex && predicate(self[index]) {
+      formIndex(after: &index)
+    }
+    return index
+  }
+}
+
+//===----------------------------------------------------------------------===//
+// startOfSuffix(while:)
+//===----------------------------------------------------------------------===//
+
+extension BidirectionalCollection {
+  /// Returns the inclusive lower bound of the suffix of elements that satisfy
+  /// the predicate.
+  ///
+  /// - Parameter predicate: A closure that takes an element of the collection
+  ///   as its argument and returns `true` if the element is part of the suffix
+  ///   or `false` if it is not. Once the predicate returns `false` it will not
+  ///   be called again.
+  ///
+  /// - Complexity: O(*n*), where *n* is the length of the collection.
+  @usableFromInline
```

Comment:
This PR looks really nice, but I think these should be inlinable rather than just usableFromInline. We want the compiler to be able to specialise them for a particular predicate closure.